# Chapter 5 -- Manipulate Arrays

## 5-1 Views and Copies

In [1]:
import numpy as np

In [2]:
mi_casa = np.array([-45, -31, -12, 0, 2, 25, 51, 99])
mi_casa

array([-45, -31, -12,   0,   2,  25,  51,  99])

In [3]:
su_casa = mi_casa

su_casa

array([-45, -31, -12,   0,   2,  25,  51,  99])

### Same or Different?

In [4]:
#reference equality
mi_casa is su_casa

True

In [5]:
id(mi_casa)

1639245697232

In [6]:
id(su_casa)

1639245697232

In [7]:
?id

In [8]:
#value inequality
mi_casa == su_casa

array([ True,  True,  True,  True,  True,  True,  True,  True])

In [9]:
su_casa[4] = 1010

In [10]:
su_casa

array([ -45,  -31,  -12,    0, 1010,   25,   51,   99])

In [11]:
mi_casa

array([ -45,  -31,  -12,    0, 1010,   25,   51,   99])

In [12]:
tree_house = np.array([-45, -31, -12, 0, 2, 25, 51, 99])

In [13]:
tree_house == mi_casa

array([ True,  True,  True,  True, False,  True,  True,  True])

In [14]:
id(tree_house)

1639245956640

In [15]:
id(mi_casa)

1639245697232

In [16]:
tree_house[0] = 214

In [17]:
tree_house

array([214, -31, -12,   0,   2,  25,  51,  99])

In [18]:
mi_casa

array([ -45,  -31,  -12,    0, 1010,   25,   51,   99])

In [19]:
tree_house == mi_casa

array([False,  True,  True,  True, False,  True,  True,  True])

In [20]:
tree_house is mi_casa

False

### View: a shallow copy

In [21]:
farm_house = tree_house.view()
farm_house.shape = (2,4)
farm_house

array([[214, -31, -12,   0],
       [  2,  25,  51,  99]])

In [22]:
tree_house

array([214, -31, -12,   0,   2,  25,  51,  99])

In [23]:
farm_house

array([[214, -31, -12,   0],
       [  2,  25,  51,  99]])

In [24]:
tree_house [3] = -111

In [25]:
farm_house

array([[ 214,  -31,  -12, -111],
       [   2,   25,   51,   99]])

### Deep Copy


In [26]:
dog_house = np.copy(tree_house)
dog_house[0] = -121
dog_house

array([-121,  -31,  -12, -111,    2,   25,   51,   99])

In [27]:
tree_house

array([ 214,  -31,  -12, -111,    2,   25,   51,   99])

In [28]:
cat_house = tree_house.copy()

In [29]:
cat_house

array([ 214,  -31,  -12, -111,    2,   25,   51,   99])

In [30]:
tree_house[0]=100

cat_house

array([ 214,  -31,  -12, -111,    2,   25,   51,   99])

## 5-2 Attributes of NumPy Arrays
- Number of dimensions
- shape
- size (number of elements)
- data type for elements
- item size for each element within an array
- type (for the array; not for elements within array)

In [31]:
import numpy as np

In [32]:
a = np.arange(24).reshape((2,3,4)) # 2 groups, 3 rows x 4 columns in each group
a

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [33]:
a.ndim

3

In [34]:
a.shape

(2, 3, 4)

In [35]:
a.size # 2 x 3 x 4 = 24

24

In [36]:
a.dtype

dtype('int32')

In [37]:
a.itemsize # the byte size of each item

4

In [38]:
type(a)

numpy.ndarray

## 5-3 Add and remove elements

- append
- horizontal stacking
- vertical stacking
- insert
- delete

In [39]:
a = np.array(np.arange(24)).reshape(2,3,4)
a

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [40]:
b=np.append(a, [5,6,7,8])
b

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23,  5,  6,  7,  8])

In [41]:
?np.append

axis : int, optional
    The axis along which `values` are appended.  If `axis` is not
    given, both `arr` and `values` are flattened before use.

In [42]:
b.shape

(28,)

In [43]:
b.reshape(7, 4)

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23],
       [ 5,  6,  7,  8]])

In [44]:
a

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [45]:
c = np.array(np.arange(24)).reshape(2,3,4) * 10 + 3
c

array([[[  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [46]:
np.append(a,c, axis=0) # axis = 0 add to group

array([[[  0,   1,   2,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23]],

       [[  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [47]:
np.append(a,c, axis=0).shape

(4, 3, 4)

In [48]:
np.append(a,c, axis=1) # axis = 1 add to rows

array([[[  0,   1,   2,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11],
        [  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23],
        [123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [49]:
np.append(a,c, axis=1).shape

(2, 6, 4)

In [50]:
np.append(a,c, axis=2) # axis = 2, add to the columns

array([[[  0,   1,   2,   3,   3,  13,  23,  33],
        [  4,   5,   6,   7,  43,  53,  63,  73],
        [  8,   9,  10,  11,  83,  93, 103, 113]],

       [[ 12,  13,  14,  15, 123, 133, 143, 153],
        [ 16,  17,  18,  19, 163, 173, 183, 193],
        [ 20,  21,  22,  23, 203, 213, 223, 233]]])

In [51]:
np.append(a,c, axis=2).shape

(2, 3, 8)

### `np.hstack()`

In [52]:
my_hay_stack = np.hstack((a,c))
my_hay_stack

array([[[  0,   1,   2,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11],
        [  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23],
        [123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [53]:
my_hay_stack[0,0,2] = 999

my_hay_stack

array([[[  0,   1, 999,   3],
        [  4,   5,   6,   7],
        [  8,   9,  10,  11],
        [  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[ 12,  13,  14,  15],
        [ 16,  17,  18,  19],
        [ 20,  21,  22,  23],
        [123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

### `np.insert()`

In [54]:
a

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [55]:
c

array([[[  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [56]:
after_insert_array = np.insert (c, 1, 444, axis=0)
after_insert_array

array([[[  3,  13,  23,  33],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[444, 444, 444, 444],
        [444, 444, 444, 444],
        [444, 444, 444, 444]],

       [[123, 133, 143, 153],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [57]:
np.insert (c, 1, 444, axis=1)

array([[[  3,  13,  23,  33],
        [444, 444, 444, 444],
        [ 43,  53,  63,  73],
        [ 83,  93, 103, 113]],

       [[123, 133, 143, 153],
        [444, 444, 444, 444],
        [163, 173, 183, 193],
        [203, 213, 223, 233]]])

In [58]:
np.insert (c, 1, 444, axis=2)

array([[[  3, 444,  13,  23,  33],
        [ 43, 444,  53,  63,  73],
        [ 83, 444,  93, 103, 113]],

       [[123, 444, 133, 143, 153],
        [163, 444, 173, 183, 193],
        [203, 444, 213, 223, 233]]])

In [59]:
?np.insert

### `np.empty()` and `np.copyto()`

In [60]:
d = np.empty(c.shape)
d

array([[[4.94065646e-324, 1.23516411e-322, 8.09880544e-312,
         8.09846071e-312],
        [8.08480396e-312, 4.94065646e-324, 9.88131292e-324,
         8.09880544e-312],
        [8.09880544e-312, 1.38338381e-322, 8.09846986e-312,
         8.08480396e-312]],

       [[8.09846986e-312, 2.12199579e-313, 8.08480396e-312,
         8.09846928e-312],
        [8.08480396e-312, 8.09846928e-312, 2.54639495e-313,
         4.94065646e-324],
        [8.09880544e-312, 8.08480396e-312, 6.95279015e-310,
         8.08480396e-312]]])

In [62]:
?np.empty

In [61]:
np.copyto(d, c)
d

array([[[  3.,  13.,  23.,  33.],
        [ 43.,  53.,  63.,  73.],
        [ 83.,  93., 103., 113.]],

       [[123., 133., 143., 153.],
        [163., 173., 183., 193.],
        [203., 213., 223., 233.]]])

In [63]:
?np.copyto

### Delete elements: `np.delete()`

In [64]:
np.delete(d, 1, axis=0)

array([[[  3.,  13.,  23.,  33.],
        [ 43.,  53.,  63.,  73.],
        [ 83.,  93., 103., 113.]]])

In [65]:
np.delete(d, 1, axis=1)

array([[[  3.,  13.,  23.,  33.],
        [ 83.,  93., 103., 113.]],

       [[123., 133., 143., 153.],
        [203., 213., 223., 233.]]])

In [66]:
np.delete(d, 1, axis=2)

array([[[  3.,  23.,  33.],
        [ 43.,  63.,  73.],
        [ 83., 103., 113.]],

       [[123., 143., 153.],
        [163., 183., 193.],
        [203., 223., 233.]]])

## 5-4 Join and split arrays

- concatenate
- stack
- split

In [67]:
import numpy as np

### concatenate
documentation: http://docs.scipy.org/doc/numpy/reference/generated/numpy.concatenate.html

In [68]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [73]:
a.shape

(2, 2)

In [74]:
b = np.array([[5, 6]])
b, b.shape

(array([[5, 6]]), (1, 2))

In [76]:
b.shape

(1, 2)

In [71]:
together = np.concatenate((a, b), axis=0) 
# add to rows, a nd b must have the same shape in columns
together

array([[1, 2],
       [3, 4],
       [5, 6]])

In [72]:
together.shape

(3, 2)

In [77]:
?np.concatenate

In [78]:
together[1,1] = 555
together

array([[  1,   2],
       [  3, 555],
       [  5,   6]])

In [79]:
a

array([[1, 2],
       [3, 4]])

In [83]:
a.shape

(2, 2)

In [80]:
c = np.array([[1, 2], [3, 4]]) *3 + 5
c

array([[ 8, 11],
       [14, 17]])

In [82]:
c.shape

(2, 2)

In [84]:
np.concatenate((a, c), axis=1)

array([[ 1,  2,  8, 11],
       [ 3,  4, 14, 17]])

### stack

In [85]:
arrays = np.zeros((5,3,4))
for n in range(5):
    arrays[n] = np.random.randn(3, 4)
    
arrays

array([[[ 1.11316294,  0.2239793 , -0.78647311, -2.3838018 ],
        [ 0.17274373,  0.38868153, -0.7329359 , -0.51985077],
        [-0.12242128, -0.90252148, -0.66442071, -0.66346813]],

       [[-0.29230368, -0.56254803, -0.36049819,  1.38053729],
        [ 0.99606415, -0.53435589,  1.1149196 ,  0.67293642],
        [-0.5379392 ,  0.68434793, -1.10038188, -2.02748775]],

       [[ 0.08117122, -0.0473436 , -0.07342641,  0.20545921],
        [-2.6698584 ,  0.69301734,  1.39977967,  0.65353335],
        [ 0.41949338,  0.24629905,  0.48783378, -0.6871976 ]],

       [[-0.60137489,  1.75423606, -0.02831208,  0.04134248],
        [-0.795815  , -0.26096506, -1.09548801, -0.50559409],
        [-0.10468468,  1.75728187, -0.02132884,  1.55126252]],

       [[ 2.35830815,  1.52761136, -0.9483501 ,  0.84264148],
        [ 1.46673622, -0.2277833 ,  0.88738174, -0.88811816],
        [-0.9861571 , -1.18861617, -0.8365403 ,  1.21872885]]])

In [87]:
stack0 = np.stack(arrays, axis=0)
stack1 = np.stack(arrays, axis=1)
stack2 = np.stack(arrays, axis=2)

In [88]:
stack0

array([[[ 1.11316294,  0.2239793 , -0.78647311, -2.3838018 ],
        [ 0.17274373,  0.38868153, -0.7329359 , -0.51985077],
        [-0.12242128, -0.90252148, -0.66442071, -0.66346813]],

       [[-0.29230368, -0.56254803, -0.36049819,  1.38053729],
        [ 0.99606415, -0.53435589,  1.1149196 ,  0.67293642],
        [-0.5379392 ,  0.68434793, -1.10038188, -2.02748775]],

       [[ 0.08117122, -0.0473436 , -0.07342641,  0.20545921],
        [-2.6698584 ,  0.69301734,  1.39977967,  0.65353335],
        [ 0.41949338,  0.24629905,  0.48783378, -0.6871976 ]],

       [[-0.60137489,  1.75423606, -0.02831208,  0.04134248],
        [-0.795815  , -0.26096506, -1.09548801, -0.50559409],
        [-0.10468468,  1.75728187, -0.02132884,  1.55126252]],

       [[ 2.35830815,  1.52761136, -0.9483501 ,  0.84264148],
        [ 1.46673622, -0.2277833 ,  0.88738174, -0.88811816],
        [-0.9861571 , -1.18861617, -0.8365403 ,  1.21872885]]])

In [89]:
stack1

array([[[ 1.11316294,  0.2239793 , -0.78647311, -2.3838018 ],
        [-0.29230368, -0.56254803, -0.36049819,  1.38053729],
        [ 0.08117122, -0.0473436 , -0.07342641,  0.20545921],
        [-0.60137489,  1.75423606, -0.02831208,  0.04134248],
        [ 2.35830815,  1.52761136, -0.9483501 ,  0.84264148]],

       [[ 0.17274373,  0.38868153, -0.7329359 , -0.51985077],
        [ 0.99606415, -0.53435589,  1.1149196 ,  0.67293642],
        [-2.6698584 ,  0.69301734,  1.39977967,  0.65353335],
        [-0.795815  , -0.26096506, -1.09548801, -0.50559409],
        [ 1.46673622, -0.2277833 ,  0.88738174, -0.88811816]],

       [[-0.12242128, -0.90252148, -0.66442071, -0.66346813],
        [-0.5379392 ,  0.68434793, -1.10038188, -2.02748775],
        [ 0.41949338,  0.24629905,  0.48783378, -0.6871976 ],
        [-0.10468468,  1.75728187, -0.02132884,  1.55126252],
        [-0.9861571 , -1.18861617, -0.8365403 ,  1.21872885]]])

In [90]:
stack2

array([[[ 1.11316294, -0.29230368,  0.08117122, -0.60137489,
          2.35830815],
        [ 0.2239793 , -0.56254803, -0.0473436 ,  1.75423606,
          1.52761136],
        [-0.78647311, -0.36049819, -0.07342641, -0.02831208,
         -0.9483501 ],
        [-2.3838018 ,  1.38053729,  0.20545921,  0.04134248,
          0.84264148]],

       [[ 0.17274373,  0.99606415, -2.6698584 , -0.795815  ,
          1.46673622],
        [ 0.38868153, -0.53435589,  0.69301734, -0.26096506,
         -0.2277833 ],
        [-0.7329359 ,  1.1149196 ,  1.39977967, -1.09548801,
          0.88738174],
        [-0.51985077,  0.67293642,  0.65353335, -0.50559409,
         -0.88811816]],

       [[-0.12242128, -0.5379392 ,  0.41949338, -0.10468468,
         -0.9861571 ],
        [-0.90252148,  0.68434793,  0.24629905,  1.75728187,
         -1.18861617],
        [-0.66442071, -1.10038188,  0.48783378, -0.02132884,
         -0.8365403 ],
        [-0.66346813, -2.02748775, -0.6871976 ,  1.55126252,
          1

### split
for information, please see: http://docs.scipy.org/doc/numpy/reference/generated/numpy.split.html

In [103]:
?np.split

In [93]:
temp = np.arange(5)
temp

array([0, 1, 2, 3, 4])

In [91]:
np.split(temp,1)

[array([0, 1, 2, 3, 4])]

In [94]:
before_split = stack0
before_split.shape

(5, 3, 4)

In [95]:
s0=np.split (before_split, 5, axis=0)
s0

[array([[[ 1.11316294,  0.2239793 , -0.78647311, -2.3838018 ],
         [ 0.17274373,  0.38868153, -0.7329359 , -0.51985077],
         [-0.12242128, -0.90252148, -0.66442071, -0.66346813]]]),
 array([[[-0.29230368, -0.56254803, -0.36049819,  1.38053729],
         [ 0.99606415, -0.53435589,  1.1149196 ,  0.67293642],
         [-0.5379392 ,  0.68434793, -1.10038188, -2.02748775]]]),
 array([[[ 0.08117122, -0.0473436 , -0.07342641,  0.20545921],
         [-2.6698584 ,  0.69301734,  1.39977967,  0.65353335],
         [ 0.41949338,  0.24629905,  0.48783378, -0.6871976 ]]]),
 array([[[-0.60137489,  1.75423606, -0.02831208,  0.04134248],
         [-0.795815  , -0.26096506, -1.09548801, -0.50559409],
         [-0.10468468,  1.75728187, -0.02132884,  1.55126252]]]),
 array([[[ 2.35830815,  1.52761136, -0.9483501 ,  0.84264148],
         [ 1.46673622, -0.2277833 ,  0.88738174, -0.88811816],
         [-0.9861571 , -1.18861617, -0.8365403 ,  1.21872885]]])]

In [96]:
type(s0)

list

In [97]:
s0[1]

array([[[-0.29230368, -0.56254803, -0.36049819,  1.38053729],
        [ 0.99606415, -0.53435589,  1.1149196 ,  0.67293642],
        [-0.5379392 ,  0.68434793, -1.10038188, -2.02748775]]])

In [99]:
s0[1].shape

(1, 3, 4)

In [100]:
s1=np.split (before_split, 3, axis=1)
s1

[array([[[ 1.11316294,  0.2239793 , -0.78647311, -2.3838018 ]],
 
        [[-0.29230368, -0.56254803, -0.36049819,  1.38053729]],
 
        [[ 0.08117122, -0.0473436 , -0.07342641,  0.20545921]],
 
        [[-0.60137489,  1.75423606, -0.02831208,  0.04134248]],
 
        [[ 2.35830815,  1.52761136, -0.9483501 ,  0.84264148]]]),
 array([[[ 0.17274373,  0.38868153, -0.7329359 , -0.51985077]],
 
        [[ 0.99606415, -0.53435589,  1.1149196 ,  0.67293642]],
 
        [[-2.6698584 ,  0.69301734,  1.39977967,  0.65353335]],
 
        [[-0.795815  , -0.26096506, -1.09548801, -0.50559409]],
 
        [[ 1.46673622, -0.2277833 ,  0.88738174, -0.88811816]]]),
 array([[[-0.12242128, -0.90252148, -0.66442071, -0.66346813]],
 
        [[-0.5379392 ,  0.68434793, -1.10038188, -2.02748775]],
 
        [[ 0.41949338,  0.24629905,  0.48783378, -0.6871976 ]],
 
        [[-0.10468468,  1.75728187, -0.02132884,  1.55126252]],
 
        [[-0.9861571 , -1.18861617, -0.8365403 ,  1.21872885]]])]

In [101]:
s1[0]

array([[[ 1.11316294,  0.2239793 , -0.78647311, -2.3838018 ]],

       [[-0.29230368, -0.56254803, -0.36049819,  1.38053729]],

       [[ 0.08117122, -0.0473436 , -0.07342641,  0.20545921]],

       [[-0.60137489,  1.75423606, -0.02831208,  0.04134248]],

       [[ 2.35830815,  1.52761136, -0.9483501 ,  0.84264148]]])

In [102]:
s2=np.split (before_split, 4, axis=2)
s2

[array([[[ 1.11316294],
         [ 0.17274373],
         [-0.12242128]],
 
        [[-0.29230368],
         [ 0.99606415],
         [-0.5379392 ]],
 
        [[ 0.08117122],
         [-2.6698584 ],
         [ 0.41949338]],
 
        [[-0.60137489],
         [-0.795815  ],
         [-0.10468468]],
 
        [[ 2.35830815],
         [ 1.46673622],
         [-0.9861571 ]]]),
 array([[[ 0.2239793 ],
         [ 0.38868153],
         [-0.90252148]],
 
        [[-0.56254803],
         [-0.53435589],
         [ 0.68434793]],
 
        [[-0.0473436 ],
         [ 0.69301734],
         [ 0.24629905]],
 
        [[ 1.75423606],
         [-0.26096506],
         [ 1.75728187]],
 
        [[ 1.52761136],
         [-0.2277833 ],
         [-1.18861617]]]),
 array([[[-0.78647311],
         [-0.7329359 ],
         [-0.66442071]],
 
        [[-0.36049819],
         [ 1.1149196 ],
         [-1.10038188]],
 
        [[-0.07342641],
         [ 1.39977967],
         [ 0.48783378]],
 
        [[-0.02831208],
 

## 5-5 Array Shape Manipulation

In [1]:
import numpy as np


### reshape
from: http://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html#numpy.reshape

NOTE: This will be a new view object if possible; otherwise, it will be a copy.

In [2]:
my_start_array = np.array(np.arange(24))
my_start_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [3]:
my_start_array.shape

(24,)

In [4]:
my_3_8_array = my_start_array.reshape((3,8))
my_3_8_array

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23]])

In [5]:
my_3_8_array.shape

(3, 8)

In [6]:
my_3_8_array[0,0] = 1234
my_3_8_array

array([[1234,    1,    2,    3,    4,    5,    6,    7],
       [   8,    9,   10,   11,   12,   13,   14,   15],
       [  16,   17,   18,   19,   20,   21,   22,   23]])

In [7]:
my_start_array

array([1234,    1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
         11,   12,   13,   14,   15,   16,   17,   18,   19,   20,   21,
         22,   23])

### `ravel()`

In [8]:
my_ravel_array = my_3_8_array.ravel()
my_ravel_array

array([1234,    1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
         11,   12,   13,   14,   15,   16,   17,   18,   19,   20,   21,
         22,   23])

In [9]:
my_start_array

array([1234,    1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
         11,   12,   13,   14,   15,   16,   17,   18,   19,   20,   21,
         22,   23])

In [10]:
my_3_8_array

array([[1234,    1,    2,    3,    4,    5,    6,    7],
       [   8,    9,   10,   11,   12,   13,   14,   15],
       [  16,   17,   18,   19,   20,   21,   22,   23]])

In [12]:
?np.ravel

### `flat` or `flatten()`

In [11]:
for n in my_3_8_array.flat:
    print(n)

1234
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


## 5-6 Rearranging Array Elements

In [13]:
import numpy as np
my_start_array = np.array(np.arange(24))
my_3_8_array = my_start_array.reshape((3,8))
my_2_3_4_array = my_3_8_array.reshape((2,3,4))

### `np.fliplr()` "flip left right"
from: http://docs.scipy.org/doc/numpy/reference/generated/numpy.fliplr.html#numpy.fliplr

Flip array in the left/right direction.

In [22]:
?np.fliplr

In [14]:
my_3_8_array

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23]])

In [15]:
np.fliplr(my_3_8_array)

array([[ 7,  6,  5,  4,  3,  2,  1,  0],
       [15, 14, 13, 12, 11, 10,  9,  8],
       [23, 22, 21, 20, 19, 18, 17, 16]])

In [16]:
my_2_3_4_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [23]:
np.fliplr(my_2_3_4_array)

array([[[ 8,  9, 10, 11],
        [ 4,  5,  6,  7],
        [ 0,  1,  2,  3]],

       [[20, 21, 22, 23],
        [16, 17, 18, 19],
        [12, 13, 14, 15]]])

### `np.flipud()` : Flip array in the up/down direction

In [19]:
my_3_8_array

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23]])

In [18]:
np.flipud(my_3_8_array)

array([[16, 17, 18, 19, 20, 21, 22, 23],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [ 0,  1,  2,  3,  4,  5,  6,  7]])

In [20]:
my_2_3_4_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [21]:
np.flipud(my_2_3_4_array)

array([[[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]],

       [[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]]])

### `np.roll()`

In [26]:
?np.roll

Signature: np.roll(a, shift, axis=None)
- Docstring: Roll array elements along a given axis.

In [24]:
my_start_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [25]:
np.roll(my_start_array, 5)

array([19, 20, 21, 22, 23,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,
       12, 13, 14, 15, 16, 17, 18])

In [27]:
np.roll(my_start_array, -5)

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
       22, 23,  0,  1,  2,  3,  4])

In [30]:
my_2_3_4_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [28]:
np.roll(my_2_3_4_array, 2)

array([[[22, 23,  0,  1],
        [ 2,  3,  4,  5],
        [ 6,  7,  8,  9]],

       [[10, 11, 12, 13],
        [14, 15, 16, 17],
        [18, 19, 20, 21]]])

In [29]:
np.roll(my_2_3_4_array, -2)

array([[[ 2,  3,  4,  5],
        [ 6,  7,  8,  9],
        [10, 11, 12, 13]],

       [[14, 15, 16, 17],
        [18, 19, 20, 21],
        [22, 23,  0,  1]]])

### `np.rot90()`

Signature: `np.rot90(m, k=1, axes=(0, 1))`
- Docstring: Rotate an array by 90 degrees in the plane specified by axes.
  Rotation direction is from the first towards the second axis.

**Parameters**

- m : array_like
    Array of two or more dimensions.
- k : integer
    Number of times the array is rotated by 90 degrees.
- axes: (2,) array_like
    The array is rotated in the plane defined by the axes.
    Axes must be different.
.. versionadded:: 1.12.0

In [31]:
my_3_8_array

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23]])

In [32]:
np.rot90(my_3_8_array)

array([[ 7, 15, 23],
       [ 6, 14, 22],
       [ 5, 13, 21],
       [ 4, 12, 20],
       [ 3, 11, 19],
       [ 2, 10, 18],
       [ 1,  9, 17],
       [ 0,  8, 16]])

In [34]:
np.rot90(my_3_8_array, k=-1)

array([[16,  8,  0],
       [17,  9,  1],
       [18, 10,  2],
       [19, 11,  3],
       [20, 12,  4],
       [21, 13,  5],
       [22, 14,  6],
       [23, 15,  7]])

In [35]:
?np.rot90

In [33]:
my_3_8_array.T

array([[ 0,  8, 16],
       [ 1,  9, 17],
       [ 2, 10, 18],
       [ 3, 11, 19],
       [ 4, 12, 20],
       [ 5, 13, 21],
       [ 6, 14, 22],
       [ 7, 15, 23]])

## 5-7 Transpose-like Operations
- transpose
- swapaxes
- rollaxes

##### Summary; from: http://stackoverflow.com/questions/15483156/generalizing-matrix-transpose-in-numpy

- Use np.transpose to permute all the axes at once. (Thanks to @jorgeca for pointing this out.)
- Use np.swapaxes to swap any two axes.
- Use np.rollaxis to "rotate" the axes.

### transpose
for more information, please see: http://docs.scipy.org/doc/numpy/reference/generated/numpy.transpose.html

**1-Dimension, no change**

In [45]:
my_start_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [46]:
np.transpose(my_start_array)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [47]:
my_start_array.T

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

##### 2-Dimensions, 
exchange rows and columns (`a[i,j] becomes a[j,i]`)

In [49]:
my_3_8_array

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23]])

In [48]:
my_3_8_array.T

array([[ 0,  8, 16],
       [ 1,  9, 17],
       [ 2, 10, 18],
       [ 3, 11, 19],
       [ 4, 12, 20],
       [ 5, 13, 21],
       [ 6, 14, 22],
       [ 7, 15, 23]])

##### 3-Dimensions or more,
use 'axes' parameter

In [50]:
my_2_3_4_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [51]:
np.transpose(my_2_3_4_array, axes=(0,2,1))

array([[[ 0,  4,  8],
        [ 1,  5,  9],
        [ 2,  6, 10],
        [ 3,  7, 11]],

       [[12, 16, 20],
        [13, 17, 21],
        [14, 18, 22],
        [15, 19, 23]]])

In [52]:
np.transpose(my_2_3_4_array, axes=(2,1,0))

array([[[ 0, 12],
        [ 4, 16],
        [ 8, 20]],

       [[ 1, 13],
        [ 5, 17],
        [ 9, 21]],

       [[ 2, 14],
        [ 6, 18],
        [10, 22]],

       [[ 3, 15],
        [ 7, 19],
        [11, 23]]])

In [53]:
?np.transpose

### swapaxes
from: http://docs.scipy.org/doc/numpy/reference/generated/numpy.swapaxes.html#numpy.swapaxes

##### returns
For Numpy >= 1.10, if a is an ndarray, then a view of a is returned; otherwise a new array is created. For earlier Numpy versions a view of a is returned only if the order of the axes is changed, otherwise the input array is returned.

In [58]:
my_2_3_4_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [59]:
np.swapaxes(my_2_3_4_array, 1, 0)

array([[[ 0,  1,  2,  3],
        [12, 13, 14, 15]],

       [[ 4,  5,  6,  7],
        [16, 17, 18, 19]],

       [[ 8,  9, 10, 11],
        [20, 21, 22, 23]]])

In [60]:
np.swapaxes(my_2_3_4_array, 1, 0).shape

(3, 2, 4)

In [61]:
np.swapaxes(my_2_3_4_array, 2, 1)

array([[[ 0,  4,  8],
        [ 1,  5,  9],
        [ 2,  6, 10],
        [ 3,  7, 11]],

       [[12, 16, 20],
        [13, 17, 21],
        [14, 18, 22],
        [15, 19, 23]]])

In [62]:
np.swapaxes(my_2_3_4_array, 2, 0)

array([[[ 0, 12],
        [ 4, 16],
        [ 8, 20]],

       [[ 1, 13],
        [ 5, 17],
        [ 9, 21]],

       [[ 2, 14],
        [ 6, 18],
        [10, 22]],

       [[ 3, 15],
        [ 7, 19],
        [11, 23]]])

### rollaxis
from: http://docs.scipy.org/doc/numpy/reference/generated/numpy.rollaxis.html#numpy.rollaxis

Roll the specified axis backwards, until it lies in a given position.

For Numpy >= 1.10 a view of a is always returned. For earlier Numpy versions a view of a is returned only if the order of the axes is changed, otherwise the input array is returned.

In [54]:
my_2_3_4_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [57]:
my_2_3_4_array.shape

(2, 3, 4)

In [55]:
np.rollaxis(my_2_3_4_array, 0, 2)

array([[[ 0,  1,  2,  3],
        [12, 13, 14, 15]],

       [[ 4,  5,  6,  7],
        [16, 17, 18, 19]],

       [[ 8,  9, 10, 11],
        [20, 21, 22, 23]]])

In [56]:
np.rollaxis(my_2_3_4_array, 0, 2).shape

(3, 2, 4)

## 5-8 Tiling arrays

In [64]:
my_start_array = np.array(np.arange(12))
my_start_array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

### tile
from: http://docs.scipy.org/doc/numpy/reference/generated/numpy.tile.html#numpy.tile

Construct an array by repeating A the number of times given by reps.

If reps has length d, the result will have dimension of max(d, A.ndim).

If A.ndim < d, A is promoted to be d-dimensional by prepending new axes. So a shape (3,) array is promoted to (1, 3) for 2-D replication, or shape (1, 1, 3) for 3-D replication. If this is not the desired behavior, promote A to d-dimensions manually before calling this function.

If A.ndim > d, reps is promoted to A.ndim by pre-pending 1’s to it. Thus for an A of shape (2, 3, 4, 5), a reps of (2, 2) is treated as (1, 1, 2, 2).

In [66]:
?np.tile

In [65]:
np.tile(my_start_array, 3)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,  0,  1,  2,  3,  4,
        5,  6,  7,  8,  9, 10, 11,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
       10, 11])

In [67]:
np.tile(my_start_array, 3).reshape((3,12))

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]])

In [69]:
my_second_array = np.array(np.arange(7))
my_second_array

array([0, 1, 2, 3, 4, 5, 6])

In [70]:
tile_1 = np.tile(my_second_array, (3, 1))
tile_1

array([[0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6]])

In [71]:
tile_2 = np.tile(tile_1, (2,2))
tile_2

array([[0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6]])

In [72]:
tile_3 = np.tile(tile_2, (3,1))
tile_3

array([[0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6],
       [0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6]])

### `np.repeat()`

In [73]:
my_second_array

array([0, 1, 2, 3, 4, 5, 6])

In [74]:
np.repeat(my_second_array, 3)

array([0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 6])

In [75]:
my_repeatable_array = np.array(np.arange(24)).reshape(2,3,4)
my_repeatable_array

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [78]:
my_repeatable_array.shape

(2, 3, 4)

In [76]:
np.repeat(my_repeatable_array, 2, axis=0)

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [79]:
np.repeat(my_repeatable_array, 2, axis=0).shape

(4, 3, 4)

In [77]:
np.repeat(my_repeatable_array, 2, axis=1)

array([[[ 0,  1,  2,  3],
        [ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [12, 13, 14, 15],
        [16, 17, 18, 19],
        [16, 17, 18, 19],
        [20, 21, 22, 23],
        [20, 21, 22, 23]]])

In [81]:
np.repeat(my_repeatable_array, 2, axis=1).shape

(2, 6, 4)

In [82]:
np.repeat(my_repeatable_array, 2, axis=2)

array([[[ 0,  0,  1,  1,  2,  2,  3,  3],
        [ 4,  4,  5,  5,  6,  6,  7,  7],
        [ 8,  8,  9,  9, 10, 10, 11, 11]],

       [[12, 12, 13, 13, 14, 14, 15, 15],
        [16, 16, 17, 17, 18, 18, 19, 19],
        [20, 20, 21, 21, 22, 22, 23, 23]]])

In [83]:
np.repeat(my_repeatable_array, 2, axis=2).shape

(2, 3, 8)